kafka-connector/src/main/java/com/google/pubsub/kafka/source/CloudPubSubSourceTask.java

```diff
@@ -133,8 +151,14 @@ public void start(Map<String, String> props) {
 
   @Override
   public List<SourceRecord> poll() throws InterruptedException {
-    ackMessages();
-    log.debug("Polling...");
+    int numberOfInflightMessages  = getInFlightMessagesCount();
+    if (numberOfInflightMessages > maxMessagesInFlight) {
+      log.warn("Number of inflight messages is to high ({} > {}). Skip polling. Sleeping for {} ms.", numberOfInflightMessages, maxMessagesInFlight, backOffPullIntervalMs);
+      Thread.sleep(backOffPullIntervalMs);
```

Comment:
The goal here is not to control memory, but to limit amount of duplicates down the stream in case ack throughput is lower than messages polling throughput. We have seen such scenarios persist for more than an hour and most of the messages produced in such cases are duplicates. In such cases it is actually the goal to get all tasks stuck until the problem resolves. 
As for you last question: kafka connect does not seem to enforce any correlation between polling and acking.